lxd/images.go

```diff
@@ -1907,5 +1912,81 @@ func imageRefresh(d *Daemon, r *http.Request) Response {
 	}
 
 	return OperationResponse(op)
+}
+
+func imageSyncBetweenNodes(d *Daemon, req api.ImagesPost, fingerprint string, imageUpload bool, file *os.File) error {
+	// Check how many nodes already have this image
+	syncNodeAddresses, err := d.cluster.ImageGetNodesHasImage(fingerprint)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get nodes for the image synchronization")
+	}
+
+	var desiredSyncNodeCount int64
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		config, err := cluster.ConfigLoad(tx)
+		if err != nil {
+			return errors.Wrap(err, "Failed to load cluster configuration")
+		}
+		desiredSyncNodeCount = config.ImageSyncNodeCount()
+		return nil
+	})
 
+	nodeCount := desiredSyncNodeCount - int64(len(syncNodeAddresses))
+	if nodeCount > 0 {
+		addresses, err := d.cluster.ImageGetNodesHasNoImage(fingerprint)
+		if err != nil {
+			return errors.Wrap(err, "Failed to get nodes for the image synchronization")
+		}
+
+		min := func(x, y int64) int64 {
+			if x > y {
+				return y
+			}
+			return x
+		}
+
+		for idx := int64(0); idx < min(int64(len(addresses)), nodeCount); idx++ {
+			client, err := cluster.Connect(addresses[idx], d.endpoints.NetworkCert(), false)
+			if err != nil {
+				return errors.Wrap(err, "Failed to connect node for image synchronization")
+			}
+
+			// We spread the image for the nodes inside of cluster and we need to double
+			// check if the image already exists since when one certain node is going to
```

Comment:
I believe that the double check mitigates the race, but does not eliminate it completely. Perhaps having and idempotent client.CreateImage API would be better. @stgraber might give more informed advice about this tho.